src/Catalyst.Core.Modules.Dfs/BlockExchange/BitSwapService.cs

```diff
@@ -0,0 +1,557 @@
+        ///   information purposes only.
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result is
+        ///   the contents of block.
+        /// </returns>
+        /// <remarks>
+        ///   Other peers are informed that the block is needed by this peer. Hopefully,
+        ///   someone will forward it to us.
+        ///   <para>
+        ///   Besides using <paramref name="cancel"/> for cancellation, the 
+        ///   <see cref="Unwant"/> method will also cancel the operation.
+        ///   </para>
+        /// </remarks>
+        public Task<IDataBlock> WantAsync(Cid id, MultiHash peer, CancellationToken cancel)
+        {
+            if (_log.IsDebugEnabled)
+            {
+                _log.Debug($"{peer} wants {id}");
+            }
+
+            var tsc = new TaskCompletionSource<IDataBlock>();
+            var want = _wants.AddOrUpdate(
+                id,
+                (key) => new WantedBlock
+                {
+                    Id = id,
+                    Consumers = new List<TaskCompletionSource<IDataBlock>> {tsc},
+                    Peers = new List<MultiHash> {peer}
+                },
+                (key, block) =>
+                {
+                    block.Peers.Add(peer);
+                    block.Consumers.Add(tsc);
+                    return block;
+                }
+            );
+
+            // If cancelled, then the block is unwanted.
+            cancel.Register(() => Unwant(id));
+
+            // If first time, tell other peers.
+            if (want.Consumers.Count != 1)
+            {
+                return tsc.Task;
+            }
+            
+            var _ = SendWantListToAllAsync(new[] {want}, full: false);
+            BlockNeeded?.Invoke(this, new CidEventArgs {Id = want.Id});
+
+            return tsc.Task;
+        }
+
+        /// <summary>
+        ///   Removes the block from the want list.
+        /// </summary>
+        /// <param name="id">
+        ///   The CID of the block to remove from the want list.
+        /// </param>
+        /// <remarks>
+        ///   Any tasks waiting for the block are cancelled.
+        ///   <para>
+        ///   No exception is thrown if the <paramref name="id"/> is not
+        ///   on the want list.
+        ///   </para>
+        /// </remarks>
+        public void Unwant(Cid id)
+        {
+            if (_log.IsDebugEnabled)
+            {
+                _log.Debug($"Unwant {id}");
+            }
+
+            if (!_wants.TryRemove(id, out var block))
+            {
+                return;
+            }
+            
+            foreach (var consumer in block.Consumers)
+            {
+                consumer.SetCanceled();
+            }
+
+            // TODO: Tell the swarm
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)